### PR TITLE
fix(lambda-nodejs): fails to find esbuild in npm workspace-local node_modules

### PR DESCRIPTION
**Pull request by kennethhli**
### Issue # (if applicable)

Closes #37545.

### Reason for this change

In npm workspaces, esbuild often ends up in a package-local `node_modules/` instead of the repo root. `PackageInstallation.detect()` used a bare `require()`, which only looks from inside `aws-cdk-lib`, so it missed those installs and fell back to Docker bundling even when local esbuild was fine.

### Description of changes

Changed `tryGetModuleVersionFromRequire()` to resolve from `process.cwd()` with `Module.createRequire()`. That matches where people usually run `cdk synth`, so workspace-local esbuild gets picked up.

### Describe any new or updated permissions being added

N/A

### Description of how you validated changes

Added unit tests in `aws-lambda-nodejs/test/util.test.ts` for workspace-local resolution and the missing-module case. Also sanity-checked the old vs new resolution behavior against the repro in #37545.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by kennethhli** [comment]
**Exemption Request**

This only changes how we detect whether esbuild is installed locally. It doesn't change the synthesized template or anything that gets deployed, so an integ snapshot wouldn't catch the bug.

The unit tests cover the actual failure mode (module present under a workspace `node_modules/`, not at the root). Happy to add an integ test if maintainers still want one.

File: packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts [hunk 1]
```diff
@@ -91,12 +92,21 @@ export function exec(cmd: string, args: string[], options?: SpawnSyncOptions) {
...
+ *
+ * @see https://github.com/aws/aws-cdk/issues/37545
  */
 export function tryGetModuleVersionFromRequire(mod: string): string | undefined {
   try {
-    // eslint-disable-next-line @typescript-eslint/no-require-imports
-    return require(`${mod}/package.json`).version;
+    // Resolve relative to the project where `cdk synth` is invoked, not from
+    // aws-cdk-lib's own node_modules path.
+    const req = createRequire(path.join(process.cwd(), 'package.json'));
```
Comment (kumvprat) [inline packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts]: This anchors module resolution at CWD only, replacing the previous anchor rather than augmenting it. 
Node's `NODE_MODULES_PATHS` only walks the *ancestor* `node_modules` of this anchor — it never descends into a nested `packages/<app>/node_modules`. In a non-hoisted npm workspace synthed from the repo root, the old bare `require()` (anchored in the nested module) walked up and resolved the workspace-local dependency; this CWD-root anchor misses it. 
Combined with the hard throw in `extractDependencies` (see L152-155 below), a previously-working `nodeModules` synth regresses to a `CannotExtractModuleVersion` abort. 


Suggest making this additive: try the CWD `createRequire` first, then fall back to a bare `require()` resolution before giving up, that keeps the #37545 fix and avoids the regression.

File: packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts [hunk 2]
```diff
@@ -91,12 +92,21 @@ export function exec(cmd: string, args: string[], options?: SpawnSyncOptions) {
...
+    return req(`${mod}/package.json`).version;
```
Comment (kumvprat) [inline packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts]: (re: L152, the version-resolution/throw path in `extractDependencies`) [nit] `extractDependencies(pkgPath, modules)` already has `pkgPath` in scope (it's passed to `tryGetModuleVersionFromPkg`). For multi-workspace monorepos where different workspaces pin different esbuild versions, anchoring the require resolution at `pkgPath` (the function's own package) rather than CWD would resolve the version the function actually gets. Edge case — worth folding in only if you're already touching this for the blocking item above.

File: packages/aws-cdk-lib/aws-lambda-nodejs/test/util.test.ts [hunk 3]
```diff
@@ -1,8 +1,9 @@
 import child_process from 'child_process';
```
Comment (kumvprat) [inline packages/aws-cdk-lib/aws-lambda-nodejs/test/util.test.ts:1]: Consider adding a test that exercises the resolution-miss / fallback path directly (a layout where the module is NOT resolvable from the CWD anchor), so the additive fallback from the blocking comment is regression-protected going forward.

**Review by kumvprat — COMMENTED** [review]
Thanks for tackling #37545

My one blocking concern is that the fix **replaces** the module-resolution anchor rather than **augmenting** it. The new anchor is `createRequire(path.join(process.cwd(), 'package.json'))` (util.ts L108) — CWD only. Node's module resolution searches only the *ancestor* `node_modules` of the anchor and never descends into a nested `packages/*/node_modules`. So for a **non-hoisted npm workspace synthed from the repo root** — one of the layouts this module must serve — the previous bare `require()` (anchored inside the nested `util.js`) walked *up* and found the workspace-local module, but the new CWD-root anchor cannot reach it. On the `nodeModules` bundling path that miss is not a graceful Docker fallback — it is a hard `throw CannotExtractModuleVersion` (util.ts L152-155), so a previously-working synth now aborts.

The fix is small and cheap: make resolution **additive** — try the CWD anchor first, then fall back to the legacy `require()` anchor before letting `extractDependencies()` throw. That keeps the #37545 fix and eliminates the regression. Details inline and a couple more nits inline to fold edge cases and test cases into this PR to make it complete.


Also **I would like to see this in action**, if you have a setup that can help verify the before and after of this change it will be good.( You can read-up here on how to test using a non-released cdk version : https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#linking)

File: packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts (hunk 1 above)
Comment (kennethhli) [reply packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts]: Updated to try cwd / pkgPath first, then fall back to the legacy `require()` so root synth doesn't regress. Added a unit test for that fallback.

Manual check:

```text
# Layout A (issue #37545): aws-cdk-lib at repo root, esbuild only in packages/my-app/node_modules
cwd: packages/my-app
  before (bare require): MISS
  after  (cwd + fallback): 0.28.0

# Layout B (reviewer case): aws-cdk-lib + esbuild both under packages/my-app, synth from repo root
cwd: repo root
  cwd-only:              MISS
  after (cwd + fallback): 0.28.0
```

File: packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts [hunk 4]
```diff
@@ -91,13 +92,36 @@ export function exec(cmd: string, args: string[], options?: SpawnSyncOptions) {
...
+ *
+ * @see https://github.com/aws/aws-cdk/issues/37545
  */
-export function tryGetModuleVersionFromRequire(mod: string): string | undefined {
+export function tryGetModuleVersionFromRequire(mod: string, fromPath?: string): string | undefined {
+  const anchors: string[] = [];
+  if (fromPath !== undefined) {
+    anchors.push(fromPath);
+  }
+  anchors.push(path.join(process.cwd(), 'package.json'));
```
Comment (kumvprat) [inline packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts:109]: Thanks for making this additive; the legacy `require()` fallback fully resolves the earlier regression. 

I have a suggestion below, it is not a hard blocker and would like to know more about this from you as you are using the nodejs lambda with bundling from a ts repo with likely multiple module versions.



***One refinement***: when `fromPath` is provided, `createRequire(fromPath)` already walks all ancestor `node_modules/` ([Node resolution algorithm](https://nodejs.org/api/modules.html#all-together)), so this unconditional `process.cwd()` anchor is redundant here and can resolve a *different* module version from CWD in a heterogeneous-version monorepo. CWD is only needed for the `PackageInstallation.detect()` path (no `fromPath`). Could we gate it?

```typescript
if (fromPath !== undefined) {
  anchors.push(fromPath);
} else {
  anchors.push(path.join(process.cwd(), 'package.json'));
}
```

The legacy `require()` fallback below the loop still covers both call paths

File: packages/aws-cdk-lib/aws-lambda-nodejs/test/util.test.ts [hunk 5]
```diff
@@ -156,6 +157,52 @@ describe('exec', () => {
   });
 });
```
Comment (kumvprat) [inline packages/aws-cdk-lib/aws-lambda-nodejs/test/util.test.ts:159]: The new tests exercise the cwd and legacy-fallback paths, but not the `fromPath` anchor that `extractDependencies` actually uses in production (`util.ts` L167). Could we add a test that passes a workspace `package.json` path as `fromPath` while `process.cwd()` points at an unrelated empty dir, asserting the `fromPath` anchor resolves? That locks in the highest-value path (and, if the `else` gating above lands, proves CWD is skipped when `fromPath` is set).

File: packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts [hunk 6]
```diff
@@ -91,13 +92,36 @@ export function exec(cmd: string, args: string[], options?: SpawnSyncOptions) {
...
+
+  for (const anchor of anchors) {
+    try {
+      const req = createRequire(anchor);
+      return req(`${mod}/package.json`).version;
```
Comment (kumvprat) [inline packages/aws-cdk-lib/aws-lambda-nodejs/lib/util.ts:114]: Users whose workspace-local esbuild was previously undetected will now silently switch from Docker to local bundling, changing the asset hash. That's the intended fix, not a blocker: but would a one-line debug annotation here(or even a console log?) recording which anchor resolved the module (or a line in README) help users who tuned CI around Docker bundling understand why their build environment changed?

This would critical for users to understand which particular path led to the resolution of the esbuild

**Review by kumvprat — COMMENTED** [review]
Thanks for your continued work on this.

I have added a few suggestions inline, the only strict requirement is the test based one.

Would like your thoughts on the other inline comments and if they dont seem useful we can avoid making those changes
